File: BitBoy/app/src/main/java/lib/observers/ObserverController.java

package lib.observers;


import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class ObserverController implements Subject {

	private Map<String,ArrayList<Observer>> observers;


	private static ObserverController instence;
	public static ObserverController shareInstence(){

		if(instence == null){

			instence = new ObserverController();
		}

		return instence;
	}

	public ObserverController() {

		Log.i("ObserverController", "ObserverController init");
		this.observers = new HashMap<String,ArrayList<Observer>>();
	}

	@Override
	public void registerObserver(Observer o, String key) {

		ArrayList<Observer> group = this.observers.get(key);
		if(group == null){

			group = new ArrayList<Observer>();
			this.observers.put(key,group);
		}
		group.add(o);

	}
	@Override
	public void removeObserver(Observer o) {

		Set<String> allkeys = this.observers.keySet();
		for (String key : allkeys) this.removeObserver(o, key);

	}
	@Override
	public void removeObserver(Observer o, String key) {
		ArrayList<Observer> group = this.observers.get(key);
		Observer observer;
		if (group != null){
			for (int i = group.size()-1 ; i >= 0; --i) {
				observer = group.get(i);
				if(observer == o) {
					group.remove(o);
				}
			}

		}

	}

	@Override
	public void notifyPost(String key,Object value, Map<String,Object> userData)
	{

		ArrayList<Observer> group = this.observers.get(key);
		if(group == null){

			return;
		}
		Observer observer;
		for (int i = 0; i < group.size(); i++) {


			observer = group.get(i);
			observer.notification(key,value, userData);
		}
	}



}

File: SMSManager/app/src/main/java/lib/observer/ObserverController.kt
package lib.observer

import android.util.Log
import java.util.ArrayList
import java.util.HashMap

open class ObserverController : Subject
{

    private val observers: MutableMap<String, ArrayList<Observer>>

    init
    {
        Log.i("ObserverController", "ObserverController init")
        this.observers = HashMap()
    }

    override fun registerObserver(o: Observer, key: String)
    {

        var group: ArrayList<Observer>? = this.observers[key]
        if (group == null)
        {
            group = ArrayList()
            this.observers[key] = group
        }
        group.add(o)
    }

    override fun removeObserver(o: Observer)
    {

        val allkeys = this.observers.keys
        for (key in allkeys) this.removeObserver(o, key)

    }

    override fun removeObserver(o: Observer, key: String)
    {
        val group = this.observers[key]
        if (group != null)
            for (observer in group) if (observer === o) group.remove(o)

    }

    override fun notifyPost(key: String)
    {
        notifyPost(key,null,null)
    }

    override fun notifyPost(key: String, value: Any?)
    {
        notifyPost(key,value,null)
    }

    override fun notifyPost(key: String, value: Any?, userData: Map<String, Any>?)
    {

        var arrayList = this.observers[key]
        val group = if (arrayList != null) arrayList else return
        for (observer in group) observer.notification(key, value, userData)
    }

    companion object
    {
        private var instence: ObserverController? = null;

        fun shareInstence(): ObserverController
        {
            if (instence == null) instence = ObserverController()
            return instence!!
        }
    }
}